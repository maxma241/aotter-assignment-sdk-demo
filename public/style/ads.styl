html,
body,
* {
  box-sizing: border-box;
}

body {
  height: 100vh;
}


.ad-container {
  position: relative;
  max-width: 1000px;
  margin: 0 auto;
  display: flex;
  flex-direction: column;
  .ad-banner {
    cursor: pointer;
    text-decoration: none;
    position: relative;
    width: 100%;
    min-height: 350px;
    margin: 10px 0px;
    padding: 0;
    .image {
      .content {
        position: relative;
        width: 100%;
        height: 100%;
        left: 0;
        margin: 0;
        padding: 0;
        img {
          &.ads-img {
            top: 0;
            left: 0;
            right: 0;
            width: 100%;
            object-fit: cover;
          }
        }
      }
    }

    .info {
      position: absolute;
      right: 42px;
      bottom: 84px;
      z-index: 3;
      padding: 0;
      margin: 0;
      &-icon {
        display: flex;
        justify-content: center;
        align-items: center;
        cursor: pointer;
        position: absolute;
        height: 28px;
        width: 28px;
        border-radius: 14px;
        border: 1px solid #fff;
        color: #000;
        background-color: #fff;
      }
    }

    .title {
      position: relative;
      display: flex;
      flex-wrap: nowrap;
      justify-content: space-between;
      flex-direction: row;
      align-items: center;
      z-index: 1;
      padding: 12px 16px;
      margin-top: -5px;
      color: #000;
      background-color: #ddd;
      overflow: hidden;
      .content {
        width: 100%;
        display: flex;
        flex-direction: column;
        .domain {
          text-transform: uppercase;
          font-weight: normal;
          color: #444;
        }
        .main {
          font-weight: 600;
        }
        span {
          text-align: left;
          max-width: 100%;
          word-break: break-word;
          text-overflow: ellipsis;
        }
      }
    }
  }


  .ad-video {
    cursor: pointer;
    text-decoration: none;
    position: relative;
    width: 100%;
    margin: 10px 0px;
    padding: 0;
    .image {
      .content {
        position: relative;
        width: 100%;
        height: 100%;
        left: 0;
        margin: 0;
        padding: 0;
        img {
          &.ads-img {
            top: 0;
            left: 0;
            right: 0;
            width: 100%;
            object-fit: cover;
          }
        }
      }
    }

    .player {
      position: absolute;
      top: 50%;
      left: 50%;
      transform: translate(-50%, -50%);
      align-items: center;
      border: 4px solid #fff;
      border-radius: 9999px;
      display: flex;
      justify-content: center;
      height: 80px;
      width: 80px;
      background-color: rgba(0, 0, 0, 0.5);;
      &-icon {
        border-color: transparent transparent transparent #fff;
        border-style: solid;
        border-width: 16px 0 16px 32px;
        height: 0;
        width: 0;
      }
    }

    .title {
      position: absolute;
      top: 0;
      left: 0;
      display: flex;
      flex-wrap: nowrap;
      justify-content: space-between;
      flex-direction: row;
      align-items: center;
      z-index: 1;
      padding: 12px 16px;
      margin-top: -5px;
      overflow: hidden;
      .content {
        width: 100%;
        display: flex;
        flex-direction: column;
        .main {
          font-size: 2vw;
          color: #fff;
          text-shadow: 1px 1px 10px #000;
        }
        span {
          text-align: left;
          max-width: 100%;
          word-break: break-word;
          text-overflow: ellipsis;
        }
      }
    }
  }
}
