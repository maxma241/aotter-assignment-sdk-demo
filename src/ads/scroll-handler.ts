

export function initialWindowScrollEventHandler(ctx: AdsContext, service: any) {
  let timeout = -1
  const stopTime = 1000
  if ('IntersectionObserver' in window ) {
    const observer = new IntersectionObserver(entries => {
      if(entries[0].isIntersecting === true) {
        timeout = setTimeout(() => {
          ctx.config.onAdImpression(ctx.currentAdData)
          service.handleImpression(ctx.currentAdData)
          // only observe once
          observer.unobserve(ctx.config.el as Element)
        }, stopTime);
      }
      else {
        window.clearTimeout(timeout)
      }
    }, { threshold: [0.5] });
    
    observer.observe(ctx.config.el as Element);
  } else {
    // for IE
    const handler = () => {
      if (isScrolledIntoView(ctx.config.el as Element) && !ctx.isInView) {
        ctx.isInView = true
        timeout = setTimeout(() => {
          if (isScrolledIntoView(ctx.config.el as Element)) {
            // call impression url once
            ctx.hasImpression = true
            ctx.config.onAdImpression(ctx.currentAdData)
            service.handleImpression(ctx.currentAdData)
          }
        }, stopTime)
      } else {
        ctx.isInView = false
        window.clearTimeout(timeout)
      }
    }
     window.addEventListener('scroll', handler)
     handler();
  }

}

function isScrolledIntoView(el: Element, percentVisible = 50) {
  let rect = el.getBoundingClientRect()
  let windowHeight = (window.innerHeight || document.documentElement.clientHeight);

  return !(
    Math.floor(100 - (((rect.top >= 0 ? 0 : rect.top) / +-(rect.height / 1)) * 100)) < percentVisible ||
    Math.floor(100 - ((rect.bottom - windowHeight) / rect.height) * 100) < percentVisible
  )
}