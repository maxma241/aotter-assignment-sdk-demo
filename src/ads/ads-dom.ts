import '../../public/style/ads.styl'


export function findAdElement(element: any) {
  if (element instanceof HTMLElement) {
    return element;
  }
  if (typeof element === 'string') {
    const el = window.document.getElementById(element)
    if (!el) {
      throw Error('ads field is not exist.')
    }
    return el;
  }
  throw Error('ads field is not exist.')
}

export function createAdsDOM(el: HTMLElement, ad: AdApiResponseVM) {
  let adDOM = ''
  if (ad.type === 'VIDEO') {
    adDOM = geVideoTemplate(ad)
  } else {
    adDOM = getBannerTemplate(ad)
  }
  el.innerHTML = adDOM
}

const geVideoTemplate = (ad: AdApiResponseVideoVM) => `
<div class="ad-container">
<a class="ad-video" taget="_blank" href="${ad.video_url}">
<div class="image">
  <div class="content">
    <img class="ads-img" src="${ad.image}" alt="" />
  </div>
</div>

<div class="player">
  <div class="player-icon"></div>
</div>
<div class="title">
  <div class="content">
    <div class="main">
      <span>${ad.title}</span>
    </div>
  </div>
</div>
</a>
</div>
`;

const getBannerTemplate = (ad: AdApiResponseBannerVM) => `
<div class="ad-container">
<a class="ad-banner" href="${ad.url}" target="_blank">
  <div class="image">
    <div class="content">
      <img class="ads-img" src="${ad.image}" alt="" />
    </div>
  </div>
  
  <div class="info">
    <div class="info-icon">
      <img src="https://static.xx.fbcdn.net/rsrc.php/v3/ys/r/SRBDu_p7126.png" alt="" height="20" width="20">
    </div>
  </div>
  <div class="title">
    <div class="content">
      <div class="domain">
        <span>agirls.aotter.net</span>
      </div>
      <div class="main">
        <span>${ad.title}</span>
      </div>
    </div>
  </div>
</a>
</div>`;