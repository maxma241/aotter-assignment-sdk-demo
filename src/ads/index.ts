// import { createEventBus, EventName, AdsEvents } from './event-handler';
import { findAdElement, createAdsDOM } from './ads-dom';
import { adsService } from '../service/ads.service';
import { initialWindowScrollEventHandler } from './scroll-handler';


export function AotterAds(config: AdsConfig) {

  const ctx: AdsContext = {
    config,
    isInView: false,
    hasImpression: false,
  }

  try {
    verifyEnv()
    config.el = findAdElement(ctx.config.el)
  } catch (error) {
    config.onAdFailed(ctx.config.el, error)
    throw error;
  }
  
  return {
    init: () => {
      initialAd(ctx)
    },
  }
}

export function verifyEnv() {
  if (!window) {
    throw Error('SDK should run on Browser environment.')
  }
}

export function initialAd(ctx: AdsContext) {
  const service = adsService(ctx.config.key);

  service.getAd(ctx.config.type)
    .then(ad => {
      // create DOM
      try {
        if (!ad) { return; }
        if (!ad.success) {
          throw Error('load ad failed.')
        }
        ctx.currentAdData = ad
        createAdsDOM(ctx.config.el as any, ctx.currentAdData)
        initialWindowScrollEventHandler(ctx, service)
        ctx.config.onAdLoaded(ctx.config.el, ad)
        return
      } catch (error) {
        // error log API
        return Promise.reject(error)
      }
    })
    .catch(err => {
      ctx.config.onAdFailed(ctx.config.el, err)
    })
}
