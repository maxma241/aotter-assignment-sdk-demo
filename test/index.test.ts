import {
  getByLabelText,
  getByText,
  getByTestId,
  queryByTestId,
  // Tip: all queries are also exposed on an object
  // called "queries" which you could import here as well
  waitFor,
  queryHelpers,
} from '@testing-library/dom'
// adds special assertions like toHaveTextContent
import '@testing-library/jest-dom/extend-expect'
import { AotterAds, initialAd } from '../src/ads/index'

function getExampleDOM() {

  const div = document.createElement('div')
  div.innerHTML = `
    <div id="ad-test"></div>
  `
  const config = {
    el: div,
    key: 'test',
    type: 'banner' as any,
    onAdLoaded(el, ad) {
      console.log(ad)
    },
    onAdFailed(el, err) {
      console.error(err)
    },
    onAdImpression(ad) {
      console.log(ad)
    },
  }
  const ad = AotterAds(config)
  return {
    ad,
    config
  }
}

describe('Ad Test', () => {

  let rootConfig: AdsConfig;

  it('Ad Root Creator Config test', async () => {
    const { ad, config } = getExampleDOM()
    rootConfig = config
    expect(ad.config).toMatchObject(config)
  })

  it('test onAdLoaded, snapshot, anchor DOM', async () => {

    const mockedService = {
      getAd: jest.fn(async (type = 'banner') => ({
        "id": "1",
        "type": "BANNER" as any,
        "title": "三星電視獨家搶先支援Apple TV App，手機還能一秒變遙控器！",
        "description": "好消息！三星在本週宣布~即日起旗下2019年全系列智慧電視、和2018年的指定機型將可透過更新支援Apple TV App以及Airplay2啦！",
        "image": "https://agirls.aotter.net/media/da724a8b-fe19-4f4e-8262-75c207ae038b.jpg",
        "url": "https://agirls.aotter.net/post/55419",
        "impression_url": "https://agirls.aotter.net?imp=1",
        "success": true
      })),
      handleImpression: jest.fn(data => data)
    }

    rootConfig.onAdLoaded = jest.fn((node, data) => data)

    const ret = await initialAd({
      config:rootConfig,
      isInView: false,
      hasImpression: false,
    }, mockedService) as AdsContext

    expect(ret.config).toMatchObject(rootConfig)
    expect(rootConfig.onAdLoaded).toBeCalled()
    expect(ret.config.el).toMatchSnapshot('banner')
    const rootAnchor = queryHelpers.queryByAttribute('class', ret.config.el as HTMLElement, 'ad-banner')
    expect(rootAnchor?.getAttribute('href')).toBe("https://agirls.aotter.net/post/55419")
  })


  it('test onAdFailed', async () => {

    const mockedService = {
      getAd: jest.fn(async (type = 'banner') => ({
        success: false,
      }) as any),
      handleImpression: jest.fn(data => data)
    }

    rootConfig.onAdFailed = jest.fn((node, data) => data)

    const ret = await initialAd({
      config:rootConfig,
      isInView: false,
      hasImpression: false,
    }, mockedService) as AdsContext

    expect(ret.config).toMatchObject(rootConfig)
    expect(rootConfig.onAdFailed).toBeCalled()
    expect(ret.config.el).toMatchSnapshot('no-ad')
  })

})