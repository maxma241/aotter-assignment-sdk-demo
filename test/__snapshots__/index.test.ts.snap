// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ad Test test initial ad and "onAdFailed": no-ad 1`] = `
<div>
  

  <div
    class="ad-container"
  >
    

    <a
      class="ad-banner"
      href="https://agirls.aotter.net/post/55419"
      target="_blank"
    >
      
  
      <div
        class="image"
      >
        
    
        <div
          class="content"
        >
          
      
          <img
            alt=""
            class="ads-img"
            src="https://agirls.aotter.net/media/da724a8b-fe19-4f4e-8262-75c207ae038b.jpg"
          />
          
    
        </div>
        
  
      </div>
      
  
  
      <div
        class="info"
      >
        
    
        <div
          class="info-icon"
        >
          
      
          <img
            alt=""
            height="20"
            src="https://static.xx.fbcdn.net/rsrc.php/v3/ys/r/SRBDu_p7126.png"
            width="20"
          />
          
    
        </div>
        
  
      </div>
      
  
      <div
        class="title"
      >
        
    
        <div
          class="content"
        >
          
      
          <div
            class="domain"
          >
            
        
            <span>
              agirls.aotter.net
            </span>
            
      
          </div>
          
      
          <div
            class="main"
          >
            
        
            <span>
              三星電視獨家搶先支援Apple TV App，手機還能一秒變遙控器！
            </span>
            
      
          </div>
          
    
        </div>
        
  
      </div>
      

    </a>
    

  </div>
</div>
`;

exports[`Ad Test test initial ad and "onAdLoaded",: banner 1`] = `
<div>
  

  <div
    class="ad-container"
  >
    

    <a
      class="ad-banner"
      href="https://agirls.aotter.net/post/55419"
      target="_blank"
    >
      
  
      <div
        class="image"
      >
        
    
        <div
          class="content"
        >
          
      
          <img
            alt=""
            class="ads-img"
            src="https://agirls.aotter.net/media/da724a8b-fe19-4f4e-8262-75c207ae038b.jpg"
          />
          
    
        </div>
        
  
      </div>
      
  
  
      <div
        class="info"
      >
        
    
        <div
          class="info-icon"
        >
          
      
          <img
            alt=""
            height="20"
            src="https://static.xx.fbcdn.net/rsrc.php/v3/ys/r/SRBDu_p7126.png"
            width="20"
          />
          
    
        </div>
        
  
      </div>
      
  
      <div
        class="title"
      >
        
    
        <div
          class="content"
        >
          
      
          <div
            class="domain"
          >
            
        
            <span>
              agirls.aotter.net
            </span>
            
      
          </div>
          
      
          <div
            class="main"
          >
            
        
            <span>
              三星電視獨家搶先支援Apple TV App，手機還能一秒變遙控器！
            </span>
            
      
          </div>
          
    
        </div>
        
  
      </div>
      

    </a>
    

  </div>
</div>
`;

exports[`Ad Test test onAdFailed: no-ad 1`] = `
<div>
  

  <div
    class="ad-container"
  >
    

    <a
      class="ad-banner"
      href="https://agirls.aotter.net/post/55419"
      target="_blank"
    >
      
  
      <div
        class="image"
      >
        
    
        <div
          class="content"
        >
          
      
          <img
            alt=""
            class="ads-img"
            src="https://agirls.aotter.net/media/da724a8b-fe19-4f4e-8262-75c207ae038b.jpg"
          />
          
    
        </div>
        
  
      </div>
      
  
  
      <div
        class="info"
      >
        
    
        <div
          class="info-icon"
        >
          
      
          <img
            alt=""
            height="20"
            src="https://static.xx.fbcdn.net/rsrc.php/v3/ys/r/SRBDu_p7126.png"
            width="20"
          />
          
    
        </div>
        
  
      </div>
      
  
      <div
        class="title"
      >
        
    
        <div
          class="content"
        >
          
      
          <div
            class="domain"
          >
            
        
            <span>
              agirls.aotter.net
            </span>
            
      
          </div>
          
      
          <div
            class="main"
          >
            
        
            <span>
              三星電視獨家搶先支援Apple TV App，手機還能一秒變遙控器！
            </span>
            
      
          </div>
          
    
        </div>
        
  
      </div>
      

    </a>
    

  </div>
</div>
`;

exports[`Ad Test test onAdLoaded, snapshot, anchor DOM: banner 1`] = `
<div>
  

  <div
    class="ad-container"
  >
    

    <a
      class="ad-banner"
      href="https://agirls.aotter.net/post/55419"
      target="_blank"
    >
      
  
      <div
        class="image"
      >
        
    
        <div
          class="content"
        >
          
      
          <img
            alt=""
            class="ads-img"
            src="https://agirls.aotter.net/media/da724a8b-fe19-4f4e-8262-75c207ae038b.jpg"
          />
          
    
        </div>
        
  
      </div>
      
  
  
      <div
        class="info"
      >
        
    
        <div
          class="info-icon"
        >
          
      
          <img
            alt=""
            height="20"
            src="https://static.xx.fbcdn.net/rsrc.php/v3/ys/r/SRBDu_p7126.png"
            width="20"
          />
          
    
        </div>
        
  
      </div>
      
  
      <div
        class="title"
      >
        
    
        <div
          class="content"
        >
          
      
          <div
            class="domain"
          >
            
        
            <span>
              agirls.aotter.net
            </span>
            
      
          </div>
          
      
          <div
            class="main"
          >
            
        
            <span>
              三星電視獨家搶先支援Apple TV App，手機還能一秒變遙控器！
            </span>
            
      
          </div>
          
    
        </div>
        
  
      </div>
      

    </a>
    

  </div>
</div>
`;
